### PR TITLE
Escalator: Method to run application as root from GUI prompt

### PR DESCRIPTION
**Pull request by ne0ndrag0n**
This PR adds a new application called "Escalator", a small utility that accepts a command to run as root and prompts the user via GUI for the user's password, similar in function to the terminal-only app `pls`. Additionally, this demonstrates Escalator's functionality by updating `NetworkSettings.af` and `Settings` to request use of Escalator when .af file indicates the Settings-category app must be run as root (as NetworkSettings requires). Escalator's icon is provided by Agni#5047 of the SerenityOS Discord (many thanks!).

### CODE REVIEW

File: Userland/Applications/Escalator/EscalatorWindow.cpp [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+    char const* shell_executable = shell.characters();
+    char const* argv[] = { shell_executable, "-c", command.characters(), nullptr };
+
+    // Escalate process privilege to root user.
+    TRY(Core::System::seteuid(0));
+    auto root_user = TRY(Core::Account::from_uid(0));
+    TRY(root_user.login());
+
+    TRY(Core::System::pledge("stdio recvfd sendfd thread cpath rpath wpath unix proc exec"));
```
Comment (ADKaster) [inline Userland/Applications/Escalator/EscalatorWindow.cpp:141]: I would expect that this pledge set can be reduced at this point in the program

File: Userland/Applications/Escalator/EscalatorWindow.cpp [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    pid_t child_pid = TRY(Core::System::posix_spawn(shell, nullptr, nullptr, const_cast<char**>(argv), environ));
+    auto waitpid_result = TRY(Core::System::waitpid(child_pid));
```
Comment (ADKaster) [inline Userland/Applications/Escalator/EscalatorWindow.cpp:143]: This looks strange. Why do we need the Escalator process to stick around as a root process and wait for the app to exit?

Can we just execve ourselves, or spawn off the process and let it be free, and shut down the GUI app entirely? This would let you avoid having to `hide()` in check_password.

File: Meta/run.sh [hunk 3]
```diff
@@ -340,7 +340,7 @@ $SERENITY_EXTRA_QEMU_ARGS
 -device isa-vga
 -device isa-ide
 $SERENITY_BOOT_DRIVE
--device i8042
```
Comment (ADKaster) [inline Meta/run.sh:343]: Could you revert these whitespace changes? If there's some formatting cleanup to do in Meta/ it should be done as a separate commit or PR.

File: Userland/Libraries/LibDesktop/AppFile.h [hunk 4]
```diff
@@ -32,6 +32,7 @@ class AppFile : public RefCounted<AppFile> {
     String icon_path() const;
     GUI::Icon icon() const;
     bool run_in_terminal() const;
+    bool requires_root() const;
```
Comment (ADKaster) [inline Userland/Libraries/LibDesktop/AppFile.h:35]: This feels like three separate  and distinct changes, but it could still be 2. The first one being updating AppFile to have this new parameter, and the second/third being updating Settings to respect the option, and marking NetworkSettings as a root-requiring process.

**Review by ADKaster — COMMENTED** [review]
Lookin pretty neato, few comments though.

File: Userland/Applications/Escalator/EscalatorWindow.cpp (hunk 1 above)
Comment (ne0ndrag0n) [reply Userland/Applications/Escalator/EscalatorWindow.cpp]: Reduced in latest push

File: Userland/Applications/Escalator/EscalatorWindow.cpp (hunk 2 above)
Comment (ne0ndrag0n) [reply Userland/Applications/Escalator/EscalatorWindow.cpp]: This more or less was imitating what was done on the Run app which served as a basis for Escalator. I think exec may be better here after all, and it's what `pls` uses. Updated.

Comment (ADKaster) [reply Userland/Applications/Escalator/EscalatorWindow.cpp]: aah, much more intuitive now. I'm still skeptical about the hide + close + show dance in some of the methods but i'll let a more GUI-minded reviewer take a look at those.

File: Userland/Applications/Escalator/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    TARGETS Escalator
+)
+
+compile_gml(Escalator.gml EscalatorGML.h escalator_gml)
+
+set(SOURCES
+    main.cpp
+    EscalatorWindow.cpp
+    EscalatorWindow.h
+    EscalatorGML.h
```
Comment (timschumi) [inline Userland/Applications/Escalator/CMakeLists.txt:13]: We usually don't put header files in `SOURCES` unless there is a reason for that.

File: Userland/Applications/Escalator/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+)
+
+serenity_app(Escalator ICON app-escalator)
```
Comment (timschumi) [inline Userland/Applications/Escalator/CMakeLists.txt:16]: At least the `32x32` icon does not appear to have been run through `optipng`, which reduces its file size from 1.29KB to 903 bytes.

File: Userland/Applications/Escalator/EscalatorWindow.cpp (hunk 2 above)
Comment (ne0ndrag0n) [reply Userland/Applications/Escalator/EscalatorWindow.cpp]: Hide/show/close is intended to center the message dialogues on the screen rather than on the password box, which imo would best match the intended Win9x visual style.

File: Userland/Applications/Escalator/EscalatorWindow.cpp [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+        }
+    }
+
+    Vector<StringView> paths = path.split_view(':');
+    for (auto const& path_entry : paths) {
+        String full_path = String::formatted("{}/{}", path_entry, m_command[0]);
+        auto file_or_error = Core::MappedFile::map(full_path);
+        if (!file_or_error.is_error())
+            return GUI::FileIconProvider::icon_for_executable(full_path);
+    }
```
Comment (timschumi) [inline Userland/Applications/Escalator/EscalatorWindow.cpp:95]: There is an existing implementation for that, `Core::File::resolve_executable_from_environment`. That also handles a `PATH` fallback, and checking if the file that we resolved is actually executable (in order to keep it consistent with what the Shell or other programs would do).

File: Userland/Applications/Escalator/main.cpp [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    args_parser.add_positional_argument(command, "Command to run at elevated privilege level", "command");
+    args_parser.parse(arguments);
+
+    TRY(Core::System::pledge("stdio recvfd sendfd thread cpath rpath wpath unix proc exec id"));
+
+    auto app = TRY(GUI::Application::try_create(arguments));
+    auto current_user = TRY(Core::Account::self());
+    auto window = TRY(EscalatorWindow::try_create(description, command, current_user));
+
+    if (current_user.uid()) {
```
Comment (timschumi) [inline Userland/Applications/Escalator/main.cpp:33]: Could probably have a `!= 0` here for clarity.

File: Userland/Applications/Escalator/EscalatorWindow.cpp [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+ErrorOr<void> EscalatorWindow::check_password()
+{
+    String password = m_password_input->text();
+    if (password.is_empty()) {
+        hide();
+        GUI::MessageBox::show_error(this, "Please enter a password."sv);
+        show();
+        return {};
```
Comment (timschumi) [inline Userland/Applications/Escalator/EscalatorWindow.cpp:107]: We can probably just leave the background window visible.

File: Userland/Applications/Escalator/EscalatorWindow.cpp [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    // FIXME: PasswordBox really should store its input directly as a SecretString.
+    Core::SecretString password_secret = Core::SecretString::take_ownership(password.to_byte_buffer());
+    if (!m_current_user.authenticate(password_secret)) {
+        hide();
+        GUI::MessageBox::show_error(this, "Incorrect or disabled password."sv);
+        m_password_input->select_all();
+        show();
```
Comment (timschumi) [inline Userland/Applications/Escalator/EscalatorWindow.cpp:116]: Same here.

**Review by timschumi — CHANGES_REQUESTED** [review]

File: Userland/Applications/Escalator/CMakeLists.txt (hunk 6 above)
Comment (ne0ndrag0n) [reply Userland/Applications/Escalator/CMakeLists.txt:14]: Updated

File: Userland/Applications/Escalator/CMakeLists.txt (hunk 5 above)
Comment (ne0ndrag0n) [reply Userland/Applications/Escalator/CMakeLists.txt]: Looks like EscalatorGML.h will still be required here, but not EscalatorWindow.h. This is so that symbol `escalator_gml` can be used when setting up the window. I will go ahead and remove EscalatorWindow.h.

File: Userland/Applications/Escalator/EscalatorWindow.cpp (hunk 9 above)
Comment (ne0ndrag0n) [reply Userland/Applications/Escalator/EscalatorWindow.cpp:107]: So if we keep the dialogue open, with the current MessageBox API it looks like we'll end up with errors that look like this:

image

MessageBox will center it on the dialog itself rather than center it on-screen. If we omit `this` in the argument, the box will center on desktop, but we lose modal functionality. RunWindow, the application I modeled much of the behaviour of Escalator after, does something similar where it hides the main window to display error dialogues in the center of the screen. 

Removing the hides and shows is trivial but, unless there's a way of manually positioning the error dialogue, it looks like it will come with that as a side effect.

Comment (ne0ndrag0n) [reply Userland/Applications/Escalator/EscalatorWindow.cpp:104]: Following discussion in discord, will go with option where error dialogue is centered on Escalator application.

File: Userland/Applications/Escalator/EscalatorWindow.cpp (hunk 7 above)
Comment (ne0ndrag0n) [reply Userland/Applications/Escalator/EscalatorWindow.cpp]: Updated

File: Userland/Applications/Escalator/EscalatorWindow.cpp [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+#include <Applications/Escalator/EscalatorGML.h>
+#include <LibCore/File.h>
+#include <LibCore/SecretString.h>
+#include <LibCore/System.h>
+#include <LibGUI/FileIconProvider.h>
+#include <LibGUI/Label.h>
+#include <LibGUI/MessageBox.h>
+#include <LibGUI/Widget.h>
+#include <stdlib.h>
+#include <unistd.h>
```
Comment (ADKaster) [inline Userland/Applications/Escalator/EscalatorWindow.cpp:19]: One last nitpick, I don't think these are used anymore thanks to the use of all the fun LibCore helpers :)

**Review by ADKaster — APPROVED** [review]
Nice work! I do have a final nitpick but this is looking like quite the idiomatic serenity app

File: Userland/Applications/Escalator/EscalatorWindow.cpp (hunk 11 above)
Comment (ne0ndrag0n) [reply Userland/Applications/Escalator/EscalatorWindow.cpp]: `environ` is used from unistd.h to carry forward the user's execution environment when calling `exec`. However, stdlib doesn't appear to be used in this file. I have removed stdlib.h as a result.

File: Userland/Applications/Escalator/EscalatorWindow.cpp [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    m_cancel_button->on_click = [this](auto) {
+        close();
+    };
+
+    m_password_input = *main_widget.find_descendant_of_type_named<GUI::PasswordBox>("password");
+}
+
+GUI::Icon EscalatorWindow::find_icon()
+{
+    auto executable_path = Core::File::resolve_executable_from_environment(m_command[0]);
```
Comment (timschumi) [inline Userland/Applications/Escalator/EscalatorWindow.cpp:68]: Very minor nitpick, but one could probably do this in advance to save on `exec` having to resolve the whole thing a second time (there should be no consistency issues because both use the same resolve implementation, but it still runs twice).

File: Userland/Applications/Escalator/EscalatorWindow.cpp [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+    if (executable_path.has_value())
+        return GUI::FileIconProvider::icon_for_executable(executable_path.value());
+
+    return GUI::Icon::default_icon("app-escalator"sv);
```
Comment (timschumi) [inline Userland/Applications/Escalator/EscalatorWindow.cpp:73]: This logic looks a bit weird. In a case where we aren't able to resolve an executable path, the `exec` will presumably fail to run anyways. The more interesting case (for the purpose of returning an icon) would likely be if there is no special icon for that executable (i.e. we'd have to specify that as a fallback for `FileIconProvider`).

**Comment by ne0ndrag0n** [comment]
Sorry to do this later in the PR process but it looked fairly important for security purposes. I have added additional functionality to require the user to pass the `-E` flag before their environment is carried up to root (minus TERM, which also mirrors `pls`). Please check latest diff.

File: Userland/Applications/Escalator/EscalatorWindow.cpp (hunk 13 above)
Comment (ne0ndrag0n) [reply Userland/Applications/Escalator/EscalatorWindow.cpp]: Updated to check once and pass to the window if valid. Error will now be thrown if the command fails to resolve and returns 127 instead of showing the window.

File: Userland/Applications/Escalator/EscalatorWindow.cpp (hunk 12 above)
Comment (ne0ndrag0n) [reply Userland/Applications/Escalator/EscalatorWindow.cpp]: Moved stuff around a bit so this is now pre-resolved before the window is even created.

File: Userland/Applications/Escalator/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+    TARGETS Escalator
+)
+
+compile_gml(Escalator.gml EscalatorGML.h escalator_gml)
+
+set(SOURCES
+    main.cpp
+    EscalatorWindow.cpp
+    EscalatorGML.h
+)
```
Comment (ADKaster) [inline Userland/Applications/Escalator/CMakeLists.txt:13]: aha! a nitpick.

CxByte pushed a [commit](https://github.com/SerenityOS/serenity/commit/166a905951023a4c183a196ef444e689fdfa9697) yesterday moving all generated headers in SOURCES into `GENERATED_SOURCES` where they belong

```suggestion
set(SOURCES
    main.cpp
    EscalatorWindow.cpp
)

set(GENERATED_SOURCES
    EscalatorGML.h
)
```

Then I think we'll be done, I was waiting on Tim to hit the approve button but this is probably been over-reviewed at this point :)

File: Userland/Applications/Escalator/EscalatorWindow.h [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+
+private:
+    EscalatorWindow(StringView executable, Vector<StringView> arguments, Options const& options);
+
+    ErrorOr<void> check_password();
+
+    Vector<StringView> m_arguments;
+    StringView m_executable;
+    Core::Account m_current_user;
+    bool m_preserve_env;
```
Comment (linusg) [inline Userland/Applications/Escalator/EscalatorWindow.h:41]: Please explicitly initialize primitives

**Review by linusg — CHANGES_REQUESTED** [review]
Please run optipng on the icons as per the contribution guidelines:

https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md#code-submission-policy

> If you have images that go along with your code, run optipng -strip all on them to optimize and strip away useless metadata - this can reduce file size from multiple kilobytes to a couple hundred bytes.

These are already quite small, but still not ideal:

> Output file size = 229 bytes (494 bytes = 68.33% decrease)

**Review by linusg — CHANGES_REQUESTED** [review]
Sorry, one more:

>  AppFile: Add support for requires_root option 

Should be

>  LibDesktop: Add support for requires_root option to AppFile

according to https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md#code-submission-policy:

> Don't use specific component names, e.g. C++ class names, as the category either - mention them in the summary instead. E.g. LibGUI: Brief description of what's being changed in FooWidget rather than FooWidget: Brief description of what's being changed

**Review by timschumi — APPROVED** [review]
